webpack/assets/javascripts/react_app/components/Layout/LayoutActions.js

```diff
@@ -0,0 +1,181 @@
+import {
+  LAYOUT_SHOW_LOADING,
+  LAYOUT_HIDE_LOADING,
+  LAYOUT_CHANGE_ACTIVE,
+  LAYOUT_RESOURCES_REQUEST,
+  LAYOUT_CHANGE_LOCATION,
+  LAYOUT_CHANGE_ORG,
+} from './LayoutConstants';
+
+export const showLoading = () => ({
+  type: LAYOUT_SHOW_LOADING,
+});
+
+export const hideLoading = () => ({
+  type: LAYOUT_HIDE_LOADING,
+});
+
+export const changeActiveMenu = primary => (dispatch) => {
+  dispatch({
+    type: LAYOUT_CHANGE_ACTIVE,
+    payload: {
+      primary,
+    },
+  });
+};
+
+export const fetchMenuItems = menuItems => (dispatch) => {
+  const activePath = window.location.pathname;
+  const items = customItems(menuItems, activePath);
+
+  dispatch({
+    type: LAYOUT_RESOURCES_REQUEST,
+    payload: {
+      items,
+    },
+  });
+};
+
+export const changeOrg = (id, org) => (dispatch) => {
+  // use ID for api
+  dispatch({
+    type: LAYOUT_CHANGE_ORG,
+    payload: {
+      org,
+    },
+  });
+};
+
+export const changeLoc = (id, location) => (dispatch) => {
+  // use ID for api
+  dispatch({
+    type: LAYOUT_CHANGE_LOCATION,
+    payload: {
+      location,
+    },
+  });
+};
+
+const customItems = (data, activePath) => {
+  const items = [];
+
+  // Menu Items
+
+  data.menu.forEach((menu) => {
+    menu.forEach((item) => {
+      let activeFlag = false;
+      const childrenArray = [];
+      item.children.forEach((child) => {
+        if (child.url === activePath) activeFlag = true;
+
+        const childObject = {
+          title: isEmpty(child.name) === true ? child.name : __(child.name),
+          isDivider: child.type === 'divider' && !isEmpty(child.name),
+          onClick: () => window.Turbolinks.visit(child.url),
```

Comment:
this should know how to opt out from turbolinks, maybe consider extracting it to a function first?

Reply:
Notice that  Katello items, and react router items should be considered as well